mlfinlab/data_structures/base_bars.py

```diff
@@ -186,7 +230,338 @@ def _get_imbalance(self, price, signed_tick, volume):
+
+        # Previous bars number of ticks and previous tick imbalances
+        self.imbalance_tick_statistics = {'num_ticks_bar': [], 'imbalance_array': []}
+
+        if analyse_thresholds is True:
+            # Array of dicts: {'timestamp': value, 'cum_theta': value, 'exp_num_ticks': value, 'exp_imbalance': value}
+            self.bars_thresholds = []
+        else:
+            self.bars_thresholds = None
+
+    def _reset_cache(self):
+        """
+        Implementation of abstract method _reset_cache for imbalance bars
+        """
+        self.open_price = None
+        self.high_price, self.low_price = -np.inf, np.inf
+        self.cum_statistics = {'cum_ticks': 0, 'cum_dollar_value': 0, 'cum_volume': 0, 'cum_buy_volume': 0}
+        self.thresholds['cum_theta'] = 0
+
+    def _extract_bars(self, data: pd.DataFrame) -> list:
+        """
+        For loop which compiles the various imbalance bars: dollar, volume, or tick.
+
+        :param data: (DataFrame) Contains 3 columns - date_time, price, and volume.
+        :return: (List) of bars built using the current batch.
+        """
+
+        # Iterate over rows
+        list_bars = []
+        for row in data.values:
+            # Set variables
+            date_time = row[0]
+            price = np.float(row[1])
+            volume = row[2]
+            dollar_value = price * volume
+            signed_tick = self._apply_tick_rule(price)
+
+            if self.open_price is None:
+                self.open_price = price
+
+            # Update high low prices
+            self.high_price, self.low_price = self._update_high_low(
+                self.high_price, self.low_price, price)
+
+            # Bar statistics calculations
+            self.cum_statistics['cum_ticks'] += 1
+            self.cum_statistics['cum_dollar_value'] += dollar_value
+            self.cum_statistics['cum_volume'] += volume
+            if signed_tick == 1:
+                self.cum_statistics['cum_buy_volume'] += volume
+
+            # Imbalance calculations
+            imbalance = self._get_imbalance(price, signed_tick, volume)
+            self.imbalance_tick_statistics['imbalance_array'].append(imbalance)
+            self.thresholds['cum_theta'] += imbalance
+
+            self.prev_price = price  # Update previous price for tick rule calculations
+
+            # Get expected imbalance for the first time, when num_ticks_init passed
+            if not list_bars and np.isnan(self.thresholds['expected_imbalance']):
+                self.thresholds['expected_imbalance'] = self._get_expected_imbalance(
+                    self.expected_imbalance_window)
+
+            if self.bars_thresholds is not None:
+                self.thresholds['timestamp'] = date_time
+                self.bars_thresholds.append(dict(self.thresholds))
+
+            self.prev_price = price  # Update previous price used for tick rule calculations
```

Comment:
1. This line is repeated.
2. The only purpose of self.prev_price is to be a price holder in self._apply_tick_rule, my suggestion would be to make this self.prev_price update a part of self._apply_tick_rule.